### PR TITLE
feat(adapters): add vanilla PostgreSQL adapter

### LINKED ISSUE #4897: New Postgres adapter: getSessionAndUser() has undefined sessionToken for email auth
### Adapter type

Custom adapter

### Environment

  System:
    OS: macOS 12.4
    CPU: (8) arm64 Apple M1
    Memory: 86.27 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.10 - /opt/homebrew/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
  Browsers:
    Chrome: 103.0.5060.114
    Firefox: 102.0.1
    Safari: 15.5
    Safari Technology Preview: 15.4
  npmPackages:
    next: ^12.2.2 => 12.2.2 
    next-auth: ^4.9.0 => 4.9.0 
    react: 17.0.2 => 17.0.2 

### Reproduction URL

https://github.com/jakecoppinger/next-auth/blob/main/packages/adapter-postgres/src/index.ts

### Describe the issue

Hello,
I'm working on a "vanilla" Postgres adapter for NextAuth (no ORM, just SQL and `pg`). I previously used Prisma and had no issues, but I'm looking to remove Prisma.

The sign in flow works up to sending an email and receiving that email. When I click on an email sign in link, the `getSessionAndUser()` method receives an undefined `sessionToken`, meaning I'm unable to find the session in the DB, resulting in the app still being signed out.

I've looked through `next-auth/packages/next-auth/src/core/routes/session.ts` and `next-auth/packages/next-auth/src/core/lib/callback-handler.ts` and I can't figure out why this would be called with an undefined value.

I didn't have any luck `yarn link`ing so I'm not sure how to find the stack trace.

Do you have any ideas of how I could debug this?
Please see the adapter at https://github.com/jakecoppinger/next-auth/blob/main/packages/adapter-postgres/src/index.ts and example SQL schema at https://github.com/jakecoppinger/next-auth/blob/main/packages/adapter-postgres/example-schema.sql

The code is based on the public gist at https://gist.github.com/gue-ni/39751d1122822ee6cf21f756b34c2c5a/revisions

Thanks very much for your help.

My adapter is as below:
```
return NextAuth(req, res, {
    adapter: PrismaAdapter(prisma),
    pages: {
      signIn: '/',
      signOut: '/auth/signout',
      verifyRequest: '/auth/verify-request',
      error: '/auth/error'
    },
    providers: [
      EmailProvider({
        server: {
          host: process.env.EMAIL_SERVER_HOST,
          port: process.env.EMAIL_SERVER_PORT,
          auth: {
            user: process.env.EMAIL_SERVER_USER,
            pass: process.env.EMAIL_SERVER_PASSWORD
          }
        },
        from: process.env.EMAIL_FROM
      }),
    ],
    secret: process.env.NEXTAUTH_SECRET
  })
```

### How to reproduce

Running `npm run dev` in a project with this adapter and schema.

Please let me know if you need a full example project to reproduce.

### Expected behavior

Should sign in and not have an undefined argument in the params.

### PR DESCRIPTION
**Pull request by jakecoppinger**
This is my first contribution to this repo - please let me know if there are things I'm missing/could improve!

All tests are passing.

## ☕️ Reasoning

What changes are being made? What feature/bug is being fixed here?

## 🧢 Checklist

- [ x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: https://github.com/nextauthjs/next-auth/issues/4897

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

File: packages/adapter-postgres/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const client = new Pool({
+  connectionString,
+});
+
+// For more information on each option (and a full list of options) go to
+// https://next-auth.js.org/configuration/options
+export default NextAuth({
+  // https://next-auth.js.org/configuration/providers
+  providers: [],
+  adapter: PostrgesAdapter(client)
```
Comment (davidmfoley) [inline packages/adapter-postgres/README.md:37]: typo (`Postrges`)

**Comment by jakecoppinger** [comment]
I'm still unable to solve a bug - `createSession` being called with an `undefined` `userId`: https://github.com/nextauthjs/next-auth/issues/4897

I initially thought I fixed it, but the issue persists. Are you able to reproduce and suggest a fix? :)

**Comment by balazsorban44** [comment]
Just remembered, could you add the adapter to these lists:

For auto-labeling issues:
https://github.com/nextauthjs/next-auth/blob/main/.github/issue-labeler.yml
https://github.com/nextauthjs/next-auth/blame/main/.github/ISSUE_TEMPLATE/3_bug_adapter.yml#L23-L34

For auto-labeling PRs
https://github.com/nextauthjs/next-auth/blob/main/.github/pr-labeler.yml

**Comment by balazsorban44** [comment]
What do you think about calling it `@next-auth/pg-adapter`? :slightly_smiling_face:

**Comment by jakecoppinger** [comment]
Renamed to `@next-auth/pg-adapter` and updated the files you mentioned :) 
I'm still stuck on this issue, any ideas? https://github.com/nextauthjs/next-auth/pull/4933#issuecomment-1186500481

**Comment by balazsorban44** [comment]
@jakecoppinger could you update the lock file so we can run tests in GitHub actions? (just run `pnpm install` and commit the result)

**Comment by craig-bowers** [comment]
Thanks for sharing your Postgres adapter! I used this as a starting point to set up NextAuth with [pg-promise](https://github.com/vitaly-t/pg-promise) instead of pg and JavaScript instead of TS.

I am running into an issue though. When I use your adapter and try signing into my Google account via [Google Provider](https://next-auth.js.org/providers/google) I'm getting a new database entry into each of my users, accounts, and sessions tables, but the next-auth.session-token cookie is not getting set when I'm redirected back to http://localhost:3000 and when I try using useSession() on the front-end the status says I'm unauthenticated.

When I bypass the adapter it lets me log in just fine and I can use useSession() to see my Google email, image, etc., and it tells me that I'm authenticated. Obviously none of the data makes its way into Postgres with it bypassed though.
` // adapter: PostgresAdapter(db),`

Any idea on what the problem might be? I can share my [...nextauth].js or adapter file if you think it'll help.

**Comment by jakecoppinger** [comment]
@craig-bowers I haven't done any testing with anything but a standard email and password unfortunately, so no ideas on my end.

I'm still stuck on an issue (#4897), did you find a way to fix that by any chance?!

**Comment by bonesoul** [comment]
found a fix now?

**Comment by bonesoul** [comment]
LF this to get merged.

**Comment by jakecoppinger** [comment]
I tried to fix my PR based on the suggestion in this comment, but I'm not having any success unfortunately: https://github.com/nextauthjs/next-auth/issues/4897#issuecomment-1363414338

`getSessionAndUser` doesn't appear to be called. I've added logging in a test project to see what gets returned from the function but it isn't being called.

Any debugging help would be appreciated!

**Comment by jakecoppinger** [comment]
Fixed! Needed to return `id` (using SQL command `RETURNING`) when creating a new user.

**Comment by jakecoppinger** [comment]
Oops, meant to close the issue and not the PR.

Issue is now fixed. Rebased and ready for review & merge!

**Comment by bonesoul** [comment]
eager now to get this merged :)

**Comment by LFCavalcanti** [comment]
Hi,

Any update to merge this PR?

**Comment by robahtou** [comment]
Any updates on this? Happy to help get this through the finish line!

File: packages/adapter-pg/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+      }
+      const originalSession: AdapterSession = result1.rows[0];
+
+      const newSession: AdapterSession = {
+        ...originalSession,
+        ...session
+      }
+      const sql = `
+        UPDATE sessions set
+        "userId" = $2, expires = $3
```
Comment (ThangHuuVu) [inline packages/adapter-pg/src/index.ts:197]: we don't need to update userId since it's never passed to this function from core.

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]
hi folks, sorry for the delay - I reviewed the PR, it looks good to me - I added a small comment. To get this PR merged, here are a couple of left-overs that we need to address:
1. Publish it as ESM-only and under `@auth/` namespace. Make sure to have correct `package.json` file and `tsconfig.json` file
2. Add documentation as JSDoc. 
3. Rebase & passing tests

You can take reference from other adapters like Kysely, Drizzle, Prisma etc.

cc @jakecoppinger @robahtou let me know if you have any questions!

**Comment by ThangHuuVu** [comment]
hi folks, I finished this PR in https://github.com/nextauthjs/next-auth/pull/8679, so I'm closing this one - thanks everyone!

**Pull request closed without merging**